docs: update the angular roadmap for v19 (#58549)
angular/angular: adev/src/content/reference/roadmap.md

### Available to experiment with

* [Incremental hydration](https://angular.dev/guide/experimental/incremental-hydration)
* [Zoneless change detection](https://angular.dev/guide/experimental/zoneless)
* [Hydration support for i18n blocks](https://angular.dev/api/platform-browser/withI18nSupport)
* Resource API

> * Effect API

Reviewer: Did you mean `afterRenderEffect`? Or is `afterRenderEffect` missing from this list?

Reviewer 2: Should we also link those items to corresponding pages?

* Linked Signal API
* Route-level render mode

### Production ready

* [Explore Angular Signals](guide/signals)
* [Event replay with SSR](https://angular.dev/api/platform-browser/withEventReplay)
* [Deferrable views](https://angular.dev/guide/defer)
* [Built-in control flow](https://angular.dev/guide/templates/control-flow)
* [Local variable declaration](https://angular.dev/guide/templates/variables)

> * [Signal inputs](https://angular.dev/guide/signals/inputs)

Reviewer: Should we mention `model` as well?

### Available to experiment with

> * [Incremental hydration](https://angular.dev/guide/experimental/incremental-hydration)

Reviewer 3: Let's use relative urls, so when on next, we stay on the same version on the doc. 
```suggestion
* [Incremental hydration](/guide/experimental/incremental-hydration)
```

* [Zoneless change detection](https://angular.dev/guide/experimental/zoneless)
* [Hydration support for i18n blocks](https://angular.dev/api/platform-browser/withI18nSupport)
* Resource API
* Effect API
* Linked Signal API

> * Route-level render mode

Reviewer 2: ```suggestion
* [Route-level render mode](https://next.angular.dev/guide/hybrid-rendering)
```

Reviewer 3: note: It's should be a relative route, as the guide update is targeting major

  In v18 we shipped experimental zoneless support in Angular. It enables developers to use the framework without including zone.js in their bundle, which improves performance, debugging experience, and interoperability. As part of the initial release we also introduced zoneless support to the Angular CDK and Angular Material.

  In v19 we introduced zoneless support in server-side rendering, addressed some edge cases, and created a schematic to scaffold zoneless projects. We transitioned <a href="https://fonts.google.com/">Google Fonts</a> to zoneless which improved performance, developer experience, and allowed us to identify gaps that we need to address before moving this feature to developer preview. Stay tuned for more updates in the next months.
  </docs-card>
  <docs-card title="Signal integrations" href="">
  We're working towards improving the integration of fundamental Angular packages, such as forms, HTTP, and router, with Signals. As part of this project, we'll seek opportunities to introduce convenient signal-based APIs or wrappers to improve the holistic developer experience.
  </docs-card>
  <docs-card title="Signal debugging in Angular DevTools" href="">

>   With the evolution of Signals in Angular, we are working on a better tooling for debugging them. High on the priority list is a UI for inspecting and debugging Signal-based components.

Reviewer 4: ```suggestion
  With the evolution of Signals in Angular, we are working on a better tooling for debugging them. High on the priority list is a UI for inspecting and debugging signal usages in components.
```

Signal-based components would connect the current impl of zone-based components to signal: true components which is not the case yet.

Author: Unsure if it'll be `signal: true`, but yeah. It should be debugging of signals overall, not signal-based components.

  </docs-card>
  <docs-card title="Signal integrations" href="">
  We're working towards improving the integration of fundamental Angular packages, such as forms, HTTP, and router, with Signals. As part of this project, we'll seek opportunities to introduce convenient signal-based APIs or wrappers to improve the holistic developer experience.
  </docs-card>
  <docs-card title="Signal debugging in Angular DevTools" href="">
  With the evolution of Signals in Angular, we are working on a better tooling for debugging them. High on the priority list is a UI for inspecting and debugging signals.
  </docs-card>
  <docs-card title="Improve HMR (Hot Module Reload)" href="https://github.com/angular/angular/issues/39367#issuecomment-1439537306">
  We're working towards faster edit/refresh cycle by enabling hot module replacement.
  

>   In Angular v19 we shipped initial support for CSS and HTML HMR. We'll continue collecting feedback to make sure we're addressing developers' needs before we mark this project as complete.

Reviewer 2: ```suggestion
  In Angular v19 we shipped initial support for CSS and template HMR. We'll continue collecting feedback to make sure we're addressing developers' needs before we mark this project as complete.
```

Reviewer 4: Should we mention that it works on external templates only?

Reviewer 2: @clydin do we support an experimental HMR for inline templates in v19?

<docs-card-container>
  <docs-card title="Modernize unit testing tooling with ng test" href="">
  In v12, we revisited the Angular end-to-end testing experience by replacing Protractor with modern alternatives such as Cypress, Nightwatch, and Webdriver.io. Next, we'd like to tackle `ng test` to modernize Angular's unit testing experience.

>   We're currently evaluating Web Test Runner, Vitest, and Jest as candidates for a new test runner for Angular projects while preserving Jasmine as assertion library to not break existing tests.

Reviewer 5: Super excited to see Vitest in this list. Thank you!!

<docs-card-container>
  <docs-card title="Modernize unit testing tooling with ng test" href="">
  In v12, we revisited the Angular end-to-end testing experience by replacing Protractor with modern alternatives such as Cypress, Nightwatch, Puppeteer, Playwright, and Webdriver.io. Next, we'd like to tackle `ng test` to modernize Angular's unit testing experience.

  We're currently evaluating Web Test Runner, Vitest, and Jest as candidates for a new test runner for Angular projects while preserving Jasmine as assertion library to not break existing tests.
  </docs-card>
  <docs-card title="Evaluating Nitro support in the Angular CLI" href="https://nitro.unjs.io/">

>   We're excited about the set of features that Nitro offers such as portability, minimal design, and file-based routing. In 2025 we'll evaluate how it fits in the Angular server-side rendering model.

Reviewer 2: ```suggestion
  We're excited about the set of features that Nitro offers such as more deployment options, improved compatibility of server-side rendering with different runtimes and file-based routing. In 2025 we'll evaluate how it fits in the Angular server-side rendering model.
```

  We'll share updates as we make progress in this investigation.
  </docs-card>
</docs-card-container>

## Fast by default

<docs-card-container>
  <docs-card title="Enable incremental hydration" href="">
  In v17 we graduated hydration from developer preview and we've been consistently observing 40-50% improvements in LCP. Since then we started prototyping incremental hydration and shared a demo on stage at ng-conf.
  

>   In v19 we shipped an experimental version of incremental hydration that you try using deferrable views. Give it a try and <a href="https://github.com/angular/angular/issues">share your feedback</a> with us!

Reviewer 2: ```suggestion
  In v19 we shipped the incremental hydration in developer preview mode, powered by `@defer` blocks. Give it a try and <a href="https://github.com/angular/angular/issues">share your feedback</a> with us!
```

  As part of this project, we'll explore the requirement space of cross framework interop and our build tooling offering to make this use case possible.
  </docs-card>
</docs-card-container>

## Completed projects

<docs-card-container>
  <docs-card title="Support two-dimensional drag-and-drop" link="Completed in Q2 2024" href="https://github.com/angular/components/issues/13372">
  As part of this project, we implemented mixed orientation support for the Angular CDK drag and drop. This is one of the repository's most highly requested features.
  </docs-card>
  <docs-card title="Event replay with SSR and prerendering" link="Completed in Q4 2024" href="https://angular.dev/api/platform-browser/withEventReplay">
  In v18 we introduced an event replay functionality when using server-side rendering or prerendering. For this feature we depend on the event dispatch primitive (previously known as jsaction) that is running on Google.com.

  In Angular v19 we graduated event replay to stable and enabled it by default for all new projects.
  </docs-card>
  <docs-card title="Integrate Angular Language Service with Schematics" link="Completed in Q4 2024" href="">

>   To make it easier for developers to use modern Angular APIs, we enabled integration between the Angular language service and schematics which allows you to refactor your app with a single click.

Reviewer 6: Is this point about something like the individual signal input functionality? I think that would be a good example to give here.

Author: We're explaining this in the upcoming v19 blog post. That's just about the integration between the two.

  </docs-card>
  <docs-card title="Streamline standalone imports with Language Service" link="Completed in Q4 2024" href="">
  As part of this initiative, the language service automatically imports components and pipes in standalone and NgModule-based apps. Additionally, to enable smaller app bundles, we'll work on allowing the language service to propose the automatic removal of unused imports.
  </docs-card>
  <docs-card title="Local template variables" link="Completed in Q3 2024" href="https://github.com/angular/angular/issues/15280">

>   Local template variables is one of the most upvoted features in the Angular issue tracker. In Q2 2024 we started initial design and prototyping. Expect updates later in 2024.

Author: Nothing more to expect about local template variables.

This section represents explorations and prototyping of potential future projects. A reasonable outcome is to decide that our current solutions are the best options. Other projects may result in RFCs, graduating to in-progress projects, or being deprioritized as the web continues to innovate along with our framework.

<docs-card-container>
  <docs-card title="Signal Forms" href="">

>   Analyze existing feedback about Angular forms and design a solution which addresses developers' requirements and uses Signals for management of reactive state.

Reviewer 2: ```suggestion
  We plan to analyze existing feedback about Angular forms and design a solution which addresses developers' requirements and uses Signals for management of reactive state.
```

  We're still in early stages of planning selectorless. We'll share a request for comments when we have an early design and we're ready for next steps.
  </docs-card>
  <docs-card title="Exploration of streamed server-side rendering" href="">
  Over the past few releases we've been working on making Angular's server-side rendering story more robust. On our priority list is to explore streamed server-side rendering for zoneless application.
  </docs-card>
  <docs-card title="Investigation for authoring format improvements" href="">
  Based on our developer surveys' results we saw there are opportunities for improving the ergonomics of the component authoring format. The first step of the process will be to gather requirements and understand the problem space in advanced to an RFC. We'll share updates as we make progress. High priority in the future work will be backward compatibility and interoperability.
  </docs-card>
  <docs-card title="Improve TestBed" href="">

>   Based on feedback over the years and the recent updates in Angular's runtime, we'll evaluate TestBed to improve developer experience and reduce boilerplate when developing unit tests.

Reviewer 2: ```suggestion
  Based on feedback over the years and the recent updates in Angular's runtime, we'll evaluate TestBed to identify opportunities to improve developer experience and reduce boilerplate when developing unit tests.
```

  As part of this project, we implemented mixed orientation support for the Angular CDK drag and drop. This is one of the repository's most highly requested features.
  </docs-card>
  <docs-card title="Event replay with SSR and prerendering" link="Completed in Q4 2024" href="https://angular.dev/api/platform-browser/withEventReplay">
  In v18 we introduced an event replay functionality when using server-side rendering or prerendering. For this feature we depend on the event dispatch primitive (previously known as jsaction) that is running on Google.com.

  In Angular v19 we graduated event replay to stable and enabled it by default for all new projects.
  </docs-card>
  <docs-card title="Integrate Angular Language Service with Schematics" link="Completed in Q4 2024" href="">
  To make it easier for developers to use modern Angular APIs, we enabled integration between the Angular language service and schematics which allows you to refactor your app with a single click.
  </docs-card>
  <docs-card title="Streamline standalone imports with Language Service" link="Completed in Q4 2024" href="">

>   As part of this initiative, the language service automatically imports components and pipes in standalone and NgModule-based apps. Additionally, to enable smaller app bundles, we'll work on allowing the language service to propose the automatic removal of unused imports.

Reviewer 2: ```suggestion
  As part of this initiative, the language service automatically imports components and pipes in standalone and NgModule-based apps. Additionally, we've added a template diagnostic to highlight unused imports in standalone components, which should help make application bundles smaller.
```

  </docs-card>
  <docs-card title="Local template variables" link="Completed in Q3 2024" href="https://github.com/angular/angular/issues/15280">

>   Local template variables is one of the most upvoted features in the Angular issue tracker. In Q2 2024 we started initial design and prototyping. Expect updates later in 2024.

Reviewer 2: ```suggestion
  We've released the support for local template variables in Angular, see [`@let` docs](https://angular.dev/api/core/@let) for additional information.
```
